Spigot-Server-Patches/0447-Fix-spawn-rates-and-add-customization-options.patch

```diff
@@ -0,0 +1,146 @@
+that a peaceful mob is everything that is not hostile and an animal is just sheep, pigs, etc. In vanilla
+all all mobs except animals spawn every tick, but Bukkit forced all peaceful mobs to spawn at the animal
+cycle, which is 400x slower than their normal tick delay per spawn, which is 1.
+
+This patch adds obfuscation helpers to EnumCreatureType to prevent further confusion and adds configuration
+options to override the Bukkit "ticks-per-spawn" and replace them by customly definable ones.
+
+Note that this bug has already been fixed in Bukkit 1.14.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+index b8131abae..079b9e43c 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -613,4 +613,27 @@ public class PaperWorldConfig {
+             log("Using improved mob spawn limits (Only Natural Spawns impact spawn limits for more natural spawns)");
+         }
+     }
++
++    public boolean overrideBukkitSpawnRates;
++    public int ticksPerMonsterSpawn;
++    public int ticksPerAnimalSpawn;
++    public int ticksPerWaterMobSpawn;
++    public int ticksPerAmbientMobSpawn;
++    private void spawnRates() {
++        overrideBukkitSpawnRates = getBoolean("override-bukkit-spawn-rates", true);
++        if (overrideBukkitSpawnRates) {
++            ticksPerMonsterSpawn = getInt("ticks-per-spawns.monster", 1);
++            ticksPerAnimalSpawn = getInt("ticks-per-spawns.animal", 400);
++            ticksPerWaterMobSpawn = getInt("ticks-per-spawns.water-mobs", 1);
++            ticksPerAmbientMobSpawn = getInt("ticks-per-spawns.ambient-mobs", 1);
++
++            log("Bukkit spawn rates are overriden and have NO EFFECT! Using paper spawn rates:");
++            log("- Monsters: " + ticksPerMonsterSpawn + " ticks per spawns");
++            log("- Animals: " + ticksPerAnimalSpawn + " ticks per spawns");
++            log("- Water Mobs: " + ticksPerWaterMobSpawn + " ticks per spawns");
++            log("- Ambient Mobs: " + ticksPerAmbientMobSpawn + " ticks per spawns");
++        } else {
++            log("Using spawn rates as defined by Bukkit. Note that spawn rates specified in paper.yml are invalid unless you turn on \"override-bukkit-spawn-rates\"!");
++        }
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/EnumCreatureType.java b/src/main/java/net/minecraft/server/EnumCreatureType.java
+index 42f6a6a93..cab4db152 100644
+--- a/src/main/java/net/minecraft/server/EnumCreatureType.java
++++ b/src/main/java/net/minecraft/server/EnumCreatureType.java
+@@ -18,6 +18,11 @@ public enum EnumCreatureType {
+ 
+     public boolean matches(Entity entity) { return innerClass().isAssignableFrom(entity.getClass()); } // Paper
+     public Class<? extends IAnimal> innerClass() { return this.a(); } // Paper - OBFHELPER
++
```

Comment:
OBFHelpers should be placed inline with what they're tryna deobf, these fields have actually changed since this patch was created

Reply:
Sorry, should be fixed now